\documentclass[]{article}

\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{hyperref}
\usepackage{natbib}

\setlength{\parindent}{0pt} %to avoid the first line indent throughout the documen

\title{Variant calling from RNA-seq data using the GATK joint genotyping workflow}
\author{Jean-Simon Brouard}

\begin{document}

\maketitle

\begin{abstract}
Put the abstract here.
\end{abstract}

\section{Introduction}
Since the introduction of RNAseq, many researchers have seen the opportunity to use this data not only for differential expression analysis, but also for calling variants \cite{Piskol2013}. Examples of such researches include a, b, c and d. Whereas trusted bioinformatic protocols exist for detecting sequence variants on a variety of DNAseq samples (germline DNA, whole-exome sequencing etc.) that come from distinct contexts \cite{Koboldt2020}, environnemental samples), protocols designed to handle RNAseq data are scarce \cite{Piskol2013}. At present the gold-standard for variant calling on RNAseq data is the GATK Per-sample workflow although an updated detailed workflow documentation for calling variants in RNAseq data is in the roadmap of the GATK experts \cite{GATK_best_RNAseq}.


Currently, researchers interested in performing GATK variant calling on RNAseq data have the option of using the fully validated Per-sample workflow \cite{GATK_RNAseq_variant_discovery} or using an in-progress advanced workflow designed for cloud computing \cite{GATK_gatk4_rnaseq_github}.
As mentionned, the Per-sample approach has several drawbacks (cite myself)

An appealing alternative would be to follow most of the GATK Best practices relative to RNAseq data and to take advantage of joint genotyping approach which is avalaible in version 3 and 4 of GATK for germline short variants and indels (ref). The joint-genotyping method has proven to be more sensitive, more flexible and to reduce computational challenges relative to the traditional calling approach \cite{GATK_jointCalling_1}. In addition, the latter approach has the advantage to facilitate the inceremental discovery of variants that origin from distinct cohorts of samples. Technically, this can be achieved by combining parts of the GATK RNAseq workflow and parts of the GATK joint genotyping workflow.

In spite that the protocol described here largely use workflows and concepts developped by the GATK team, the reader should be aware that it has not been validated by GATK experts. We have shown previously that a similar workflow has and we did not ahve experienced any problesm (Brouard)

Here we present GATK4 and fully updated version of this approach.

an update end-to-end analysis of...



Figure 1 present the workflow proposed here.




% figure 1 
\input{figures/fig1}




In the next section, we describe how the diverse programs required to perform the whole analysis can be installed.

\include{Materials}

\include{Methods}

\bibliographystyle{unsrtnat}
\bibliography{bibliography/RNAseq_VC}





\end{document}
